lib/surface/compiler/ast_translator.ex

```diff
@@ -0,0 +1,281 @@
+        name: name
+      }
+    else
+      raise_unexpected_subblock_error!(name, parent, meta)
+    end
+  end
+
+  def context_for_subblock(%{tags: []}, name, _, node_meta) do
+    raise Parser.parse_error("<#{name}> is not allowed at the root of the template", node_meta)
+  end
+
+  def context_for_subblock(state, name, parent, node_meta) do
+    raise_subblock_parent_not_a_construct_error!(state, name, parent, node_meta)
+  end
+
+  defp node_type_and_alias(<<"#", first, _rest::binary>> = name, _meta) when first in ?A..?Z,
+    do: {AST.MacroComponent, String.slice(name, 1..-1)}
+
+  defp node_type_and_alias(<<"#", first, _rest::binary>> = name, _meta) when first in ?a..?z,
+    do: {Surface.Construct, String.slice(name, 1..-1)}
+
+  defp node_type_and_alias(name, _meta) when name in ["template", "slot"],
+    do: {Surface.Construct, name}
+
+  defp node_type_and_alias(<<first, _::binary>> = name, _meta) when first in ?A..?Z,
+    do: {AST.Component, name}
+
+  defp node_type_and_alias(name, %{void_tag?: true}), do: {AST.VoidTag, name}
+  defp node_type_and_alias(name, _meta), do: {AST.Tag, name}
+
+  defp module_for_node(_state, Surface.Construct, name, meta) do
+    if module = @constructs[name] do
+      module
+    else
+      raise_unknown_construct_error!(name, meta)
+    end
+  end
+
+  defp module_for_node(state, type, name, _meta)
+       when type in [AST.Component, AST.MacroComponent] do
+    Helpers.actual_component_module!(name, state.caller)
+  end
+
+  defp module_for_node(_state, _type, _name, _meta) do
+    nil
+  end
+
+  defp attribute_type_and_opts(context, name, meta)
+       when is_component(context) or is_tag(context) do
+    Surface.TypeHandler.attribute_type_and_opts(context.module, name, meta)
+  end
+
+  defp attribute_type_and_opts(context, name, meta)
+       when is_construct(context) or is_subblock(context) do
+    block_name = if is_subblock(context), do: context.name, else: :default
+    type = context.module.attribute_type(block_name, name, meta)
+    {type, []}
+  end
+
+  defp parse_value(state, type, name, {:expr, value, expr_meta}, _attr_meta) do
+    meta = to_meta(state, expr_meta)
+
+    %AST.AttributeExpr{
+      original: value,
+      value: Surface.TypeHandler.expr_to_quoted!(value, name, type, meta),
+      meta: meta
+    }
+  end
+
+  defp parse_value(_state, type, name, value, meta) do
+    Surface.TypeHandler.literal_to_ast_node!(type, name, value, meta)
+  end
+
+  defp to_meta(state, parse_meta, extra \\ []) do
+    %AST.Meta{
+      line: parse_meta.line,
+      column: parse_meta.column,
+      file: parse_meta.file,
+      caller: state.caller,
+      checks: state.checks,
+      node_alias: extra[:node_alias],
+      module: extra[:module]
+    }
+  end
+
+  defp raise_default_subblock_not_allowed_error!(parent, meta) do
+    message = """
+    #{parent.name} requires all content to be wrapped in one of the following \
+    sub blocks: #{Enum.join(parent.allowed_subblocks)}
```

Comment:
We also be able to give a hint depending on the parent:

Case:
```
Hint: case construct should be used like
<#case {@var}>
  <#match {[first | _ ]}>
    ...
  <#match {_}>
    ...
</#case>
```

Reply:
that's a good idea - it might make sense to turn the `valid_subblocks` callback into a `validate_subblock` or similar to enable construct specific messages without this having to know anything specific about that construct